src/main/java/de/rwth/idsg/steve/config/OcppConfiguration.java

```diff
@@ -52,46 +44,55 @@
+
+    @Bean
+    public EndpointImpl ocpp12Endpoint() {
+        return createDefaultEndpoint(ocpp12Server, "/CentralSystemServiceOCPP12");
     }
 
-    private void configure(Bus bus) {
-        List<Interceptor<? extends Message>> interceptors = asList(new MessageIdInterceptor(), messageHeaderInterceptor);
-        List<Feature> logging = singletonList(LoggingFeatureProxy.INSTANCE.get());
+    @Bean
+    public EndpointImpl ocpp15Endpoint() {
+        return createDefaultEndpoint(ocpp15Server, "/CentralSystemServiceOCPP15");
+    }
+
+    @Bean
+    public EndpointImpl ocpp16Endpoint() {
+        return createDefaultEndpoint(ocpp16Server, "/CentralSystemServiceOCPP16");
+    }
 
-        createOcppService(bus, ocpp12Server, "/CentralSystemServiceOCPP12", interceptors, logging);
-        createOcppService(bus, ocpp15Server, "/CentralSystemServiceOCPP15", interceptors, logging);
-        createOcppService(bus, ocpp16Server, "/CentralSystemServiceOCPP16", interceptors, logging);
+    /**
+     * Just a dummy service to route incoming messages to the appropriate service version.
+     */
+    @Bean
+    public EndpointImpl routerEndpoint(EndpointImpl ocpp12Endpoint,
+                                       EndpointImpl ocpp15Endpoint,
+                                       EndpointImpl ocpp16Endpoint) {
+        var mediator = new MediatorInInterceptor(List.of(ocpp12Endpoint, ocpp15Endpoint, ocpp16Endpoint));
+        return createEndpoint(
+            ocpp12Server, SteveProperties.ROUTER_ENDPOINT_PATH, List.of(mediator), Collections.emptyList()
+        );
+    }
 
-        // Just a dummy service to route incoming messages to the appropriate service version. This should be the last
-        // one to be created, since in MediatorInInterceptor we go over created/registered services and build a map.
-        //
-        List<Interceptor<? extends Message>> mediator = singletonList(new MediatorInInterceptor(bus));
-        createOcppService(bus, ocpp12Server, CONFIG.getRouterEndpointPath(), mediator, Collections.emptyList());
+    private EndpointImpl createDefaultEndpoint(Object serviceBean, String address) {
+        return createEndpoint(
+            serviceBean, address, List.of(messageIdInterceptor, messageHeaderInterceptor), List.of(LoggingFeatureProxy.INSTANCE.get())
+        );
     }
 
-    private void createOcppService(Bus bus, Object serviceBean, String address,
-                                   List<Interceptor<? extends Message>> interceptors,
-                                   Collection<? extends Feature> features) {
-        JaxWsServerFactoryBean f = new JaxWsServerFactoryBean();
-        f.setBus(bus);
-        f.setServiceBean(serviceBean);
-        f.setAddress(address);
-        f.getFeatures().addAll(features);
-        f.getInInterceptors().addAll(interceptors);
-        f.create();
+    private EndpointImpl createEndpoint(Object serviceBean, String address,
```

Comment:
Should return `Endpoint`

Reply:
need EndpointImpl in order to have access to the server it encapsulates for the logic in MediatorInInterceptor